frontend/src/App.js

```diff
@@ -0,0 +1,103 @@
+import React from 'react';
+import {
+    BrowserRouter as Router,
+    Switch,
+    Route,
+    Redirect
+} from 'react-router-dom';
+import { StateContext } from './State';
+
+import LoginView from './views/LoginView';
+import TicketsView from './views/TicketsView';
+import SubmitView from './views/SubmitView';
+import ProfileView from './views/ProfileView';
+import WorkcenterView from './views/WorkcenterView';
+import NavigationBar from './components/NavigationBar';
+
+const App = () => {
+    const context = React.useContext(StateContext);
+
+    return (
+        <Router>
+            <Switch>
+                <Route exact path='/'>
+                    {
+                        context.user.id !== 0
+                        ? <Redirect to='/tickets' />
+                        : <Redirect to='/login' />
+                    }
+                </Route>
+
+                <Route path='/login'>
+                    {
+                        context.user.id !== 0
+                        ? <Redirect to='/profile' />
+                        : <LoginView />
+                    }
+                </Route>
+
+                <Route exact path='/tickets'>
+                    {
+                        context.user.id !== undefined
+                        ? (
+                            <div id="wrapper">
+                                <NavigationBar />
+                                <div id="content">
+                                    <TicketsView />
+                                </div>
+                            </div>
+                        )
+                        : <Redirect to='/login' />
+                    }
+                </Route>
+
+                <Route exact path='/tickets/submit'>
+                    {
+                        context.user.id !== undefined
+                        ? (
+                            <div id="wrapper">
+                                <NavigationBar />
+                                <div id="content">
+                                    <SubmitView />
+                                </div>
+                            </div>
+                        )
+                        : <Redirect to='/login' />
```

Comment:
This pattern gets repeated several times within App.js. While some of what these ternaries set is determined by the route, the actual operation of checking if `context.user.id !== undefined` and redirecting to login if necessary is probably something we could re-work as a single function. There is a principle in DRY programming (Don't Repeat Yourself) that if we find a pattern that we need to re-create at least 3 times, we should try to find a way to encapsulate it separately - I think this is a possible candidate for that process.